File: Java/src/ToyORB/messages/ServiceRegistrationMessage.java
package ToyORB.messages;

import ToyORB.NetUtils;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.Objects;

public final class ServiceRegistrationMessage extends Message {
	public static final MessageType MESSAGE_TYPE = MessageType.NAME_REGISTER;

	private final String serviceName;
	private final String serviceType;
	private final int port;

	public ServiceRegistrationMessage(String serviceName, String serviceType, int port) {
		super(MESSAGE_TYPE);
		this.serviceName = Objects.requireNonNull(serviceName);
		this.serviceType = Objects.requireNonNull(serviceType);
		this.port = port;
	}

	public String getServiceName() {
		return serviceName;
	}

	public String getServiceType() {
		return serviceType;
	}

	public int getPort() {
		return port;
	}

	@Override
	protected void onWriteInto(DataOutputStream stream) throws IOException {
		NetUtils.writeStringInto(stream, serviceName);
		NetUtils.writeStringInto(stream, serviceType);
		stream.writeInt(port);
	}

	@Override
	public String toString() {
		return "ServerRegistrationMessage{" + serviceName + ", " + serviceType + ", " + port + "}";
	}

	@SuppressWarnings("unused")
	protected static final MessageReader<ServiceRegistrationMessage> READER = new MessageReader<ServiceRegistrationMessage>(MESSAGE_TYPE) {
		@Override
		public ServiceRegistrationMessage readMessageFrom(DataInputStream stream) throws IOException {
			String serviceName = NetUtils.readStringFrom(stream);
			String serviceType = NetUtils.readStringFrom(stream);
			int port = stream.readInt();
			return new ServiceRegistrationMessage(serviceName, serviceType, port);
		}
	};
}

File: .NET Core/ToyORB/Messages/ServiceRegistrationMessage.cs
using System;
using System.IO;

namespace ToyORB.Messages
{
    public class ServiceRegistrationMessage : Message
    {
        public static readonly MessageType MESSAGE_TYPE = MessageType.NameRegister;

        public string ServiceName { get; }
        public string ServiceType { get; }
        public int Port { get; }

        public ServiceRegistrationMessage(string serviceName, string serviceType, int port) : base(MESSAGE_TYPE)
        {
            ServiceName = serviceName ?? throw new ArgumentNullException(nameof(serviceName));
            ServiceType = serviceType ?? throw new ArgumentNullException(nameof(serviceType));
            Port = port;
        }

        public override void WriteInto(BinaryWriter stream)
        {
            base.WriteInto(stream);
            stream.WriteBigEndian(ServiceName);
            stream.WriteBigEndian(ServiceType);
            stream.WriteBigEndian(Port);
        }

        public override string ToString()
        {
            return $"ServiceRegistrationMessage{{{ServiceName}, {ServiceType}, {Port}}}";
        }

        protected internal class MessageReader : MessageReader<ServiceRegistrationMessage>
        {
            public MessageReader() : base(MESSAGE_TYPE) { }

            public override ServiceRegistrationMessage ReadMessageFrom(BinaryReader stream)
            {
                string serviceName = stream.ReadStringBigEndian();
                string serviceType = stream.ReadStringBigEndian();
                int port = stream.ReadInt32BigEndian();
                return new ServiceRegistrationMessage(serviceName, serviceType, port);
            }
        }

        protected internal static MessageReader<ServiceRegistrationMessage> Reader = new MessageReader();
    }
}